gxmail/src/main/java/com/genexus/internet/POP3SessionJavaMail.java

```diff
@@ -84,339 +78,283 @@ public void login(GXPOP3Session sessionInfo)
+				foundOne = true;
+			else
+				i++;
+		}
+		return foundOne;
+	}
+
+	private void handlePart(Part part, GXMailMessage gxmessage) throws MessagingException, IOException {
+		String disposition = part.getDisposition();
+
+		boolean isXForwardedFor = part.getContent() instanceof MimeMessage && // Para soportar attachments de mails que llegan a la casilla con el header "X-Forwarded-For"
+			findContentTypeHeaderStartingWithMessage(part);
+
+		if (System.getProperties().getProperty("DownloadAllMailsAttachment", null) != null && isXForwardedFor)
+			handlePart((MimeMessage) part.getContent(), gxmessage);
+
+		if (part.isMimeType("text/plain")) {
+			gxmessage.setText(part.getContent().toString());
+		}
+		if (part.isMimeType("text/html")) {
+			gxmessage.setHtmltext(part.getContent().toString());
+		}
+		if (part.isMimeType("multipart/*")) {
+			handleMultipart((Multipart) part.getContent(), gxmessage);
+		}
+		if (disposition == null && part.isMimeType("application/*")) {
+			disposition = "UNKNOWN";
+		}
+		if (this.downloadAttachments && (disposition != null && (disposition.equalsIgnoreCase(Part.ATTACHMENT) || disposition.equalsIgnoreCase(Part.INLINE) || disposition.equalsIgnoreCase("UNKNOWN")))) {
+			String fileName = "";
+			if (part.getFileName() != null)
+				fileName = MimeUtility.decodeText(part.getFileName());
+			else if (isXForwardedFor && ((MimeMessage) part.getContent()).getFileName() != null)
+				fileName = MimeUtility.decodeText(((MimeMessage) part.getContent()).getFileName());
+			else if (!(part.getContent() instanceof MimeMessage) || ((MimeMessage) part.getContent()).getFileName() == null)
+				fileName = SpecificImplementation.GXutil.getTempFileName("tmp");
+
+			String cid = getAttachmentContentId(part);
+			if (disposition.equalsIgnoreCase(Part.INLINE) && !cid.isEmpty()) {
+				fileName = String.format("%s_%s", cid, fileName);
+				String newHTML = gxmessage.getHtmltext().replace(cid, fileName);
+				gxmessage.setHtmltext(newHTML);
+			}
+
+			try (InputStream is = part.getContent() instanceof MimeMessage ? ((MimeMessage) part.getContent()).getInputStream() : part.getInputStream()) {
+				saveFile(fileName, is);
+			}
+			attachs.add(attachmentsPath + fileName);
+			gxmessage.setAttachments(attachs);
 		}
+	}
+
+	private String getAttachmentContentId(Part part) throws MessagingException {
+		String cid = "";
+		String[] cids = part.getHeader("Content-ID");
+		if (cids != null && !cids[0].isEmpty() && cids[0].startsWith("<") && cids[0].endsWith(">")) {
+			cid = cids[0].substring(1, cids[0].length() - 1);
+		}
+		return cid;
+	}
+
+	private void saveFile(String filename, InputStream input) throws IOException {
+		filename = filename.replace("/", "_")
+			.replace("\\", "_")
+			.replace(":", "_")
+			.replace("*", "_")
+			.replace("?", "_")
+			.replace("\"", "_")
+			.replace("<", "_")
+			.replace(">", "_")
+			.replace("|", "_")
+			.replace("\n", "_")
+            .replace("\r", "_");
```

Comment:
It would be nice if you use Pattern to replace file name, like this:

private static final String FILE_INVALID_CHARS = "[\\\\/:*?\"<>|\\n\\r]";
private static final Pattern FILE_INVALID_PATTERN = Pattern.compile(FILE_INVALID_CHARS);
filename = FILE_INVALID_PATTERN.matcher(filename).replaceAll("_");